### PR TITLE
Update AWS SDK V2 : es-discovery-ec2

### PR DESCRIPTION
**Pull request by jeeminso**
## Summary of the changes / Why this improves CrateDB
AWS SDK upgrade to V2 for es-discovery-ec2. Referred to https://github.com/aws/aws-sdk-java-v2/blob/master/docs/LaunchChangelog.md.

## Checklist

 - [ ] Added an entry in the latest `docs/appendices/release-notes/<x.y.0>.rst` for user facing changes
 - [ ] Updated documentation & `sql_features` table for user facing changes
 - [ ] Touched code is covered by tests
 - [ ] [CLA](https://crate.io/community/contribute/cla/) is signed
 - [ ] This does not contain breaking changes, or if it does:
    - It is released within a major release
    - It is recorded in the latest `docs/appendices/release-notes/<x.y.0>.rst`
    - It was marked as deprecated in an earlier release if possible
    - You've thought about the consequences and other components are adapted
      (E.g. AdminUI)

### CODE REVIEW

File: plugins/es-discovery-ec2/src/main/java/org/elasticsearch/discovery/ec2/AwsEc2ServiceImpl.java [hunk 1]
```diff
@@ -50,68 +52,55 @@ class AwsEc2ServiceImpl implements AwsEc2Service {
...
-        final RetryPolicy retryPolicy = new RetryPolicy(
-            RetryPolicy.RetryCondition.NO_RETRY_CONDITION,
-            (originalRequest, exception, retriesAttempted) -> {
-                // with 10 retries the max delay time is 320s/320000ms (10 * 2^5 * 1 * 1000)
-                LOGGER.warn("EC2 API request failed, retry again. Reason was:", exception);
-                return 1000L * (long) (10d * Math.pow(2, retriesAttempted / 2.0d) * (1.0d + rand.nextDouble()));
-            },
-            10,
-            false);
-        clientConfiguration.setRetryPolicy(retryPolicy);
```
Comment (jeeminso) [inline plugins/es-discovery-ec2/src/main/java/org/elasticsearch/discovery/ec2/AwsEc2ServiceImpl.java:101]: Couldn't find `v2` equivalent but there are different back off strategies (`FullJitterBackoffStrategy`, `FixedDelayBackoffStrategy`, or `EqualJitterBackoffStrategy`) we can use instead?

**Comment by jeeminso** [comment]
Hi @seut @matriv do you have any suggestions for https://github.com/crate/crate/pull/14973#discussion_r1374510553 and https://github.com/crate/crate/pull/14973#discussion_r1374512509? Since they are ported from ES I was thinking we could be a bit lenient. Also there are no tests to verify these changes right?

File: plugins/es-discovery-ec2/pom.xml [hunk 2]
```diff
@@ -34,9 +34,16 @@
     </dependency>
 
     <dependency>
-      <groupId>com.amazonaws</groupId>
-      <artifactId>aws-java-sdk-ec2</artifactId>
-      <version>${versions.aws}</version>
+      <groupId>software.amazon.awssdk</groupId>
+      <artifactId>ec2</artifactId>
+      <version>2.21.8</version>
+    </dependency>
```
Comment (jeeminso) [inline plugins/es-discovery-ec2/pom.xml:40]: will be reverted when merging to master

File: plugins/es-discovery-ec2/pom.xml [hunk 3]
```diff
@@ -34,9 +34,16 @@
...
+
+    <dependency>
+      <groupId>software.amazon.awssdk</groupId>
+      <artifactId>apache-client</artifactId>
+      <version>2.21.8</version>
+      <scope>compile</scope>
```
Comment (jeeminso) [inline plugins/es-discovery-ec2/pom.xml:46]: without this the following import will cause compile error `import software.amazon.awssdk.http.apache.ApacheHttpClient;`

File: plugins/es-discovery-ec2/src/main/java/org/elasticsearch/discovery/ec2/AwsEc2ServiceImpl.java (hunk 1 above)
Comment (seut) [reply plugins/es-discovery-ec2/src/main/java/org/elasticsearch/discovery/ec2/AwsEc2ServiceImpl.java:101]: Looks like still available? https://sdk.amazonaws.com/java/api/latest/software/amazon/awssdk/core/retry/RetryPolicy.html

**Comment by seut** [comment]
> Since they are ported from ES I was thinking we could be a bit lenient. 

What exactly can be more lenient?

> Also there are no tests to verify these changes right?

Right, not sure why. Could you add some?

**Comment by jeeminso** [comment]
> What exactly can be more lenient?

I meant to be lenient in terms of using a standard retry policy provided by AWS instead since I could not figure out how to port the existing one.

> Is this still in a draft mode? What is missing otherwise or what should I review?

There is nothing missing, but I will add more tests before turning this into a non-draft.

File: plugins/es-discovery-ec2/src/main/java/org/elasticsearch/discovery/ec2/AwsEc2ServiceImpl.java [hunk 4]
```diff
@@ -50,68 +52,55 @@ class AwsEc2ServiceImpl implements AwsEc2Service {
...
-        return client;
-    }
 
-    // pkg private for tests
-    static ClientConfiguration buildConfiguration(Ec2ClientSettings clientSettings) {
-        final ClientConfiguration clientConfiguration = new ClientConfiguration();
+        // NOT supported by AWS SDK V2
         // the response metadata cache is only there for diagnostics purposes,
         // but can force objects from every response to the old generation.
-        clientConfiguration.setResponseMetadataCacheSize(0);
```
[the remark this answers is not in the document]
Comment (seut) [reply plugins/es-discovery-ec2/src/main/java/org/elasticsearch/discovery/ec2/AwsEc2ServiceImpl.java:81]: Looks to me that this is no longer needed.

**Comment by jeeminso** [comment]
on hold until next cycle

**Pull request closed without merging**